File: python/segmenter/page_splitter.py
"""Split TTA's single-variable OCR string into ordered per-page texts.

TTA emits the whole document as ONE OCR string with page separators it inserts,
e.g. `[PAGE 1]`, `[PAGE 2]`, ... The hard part: a document's own body can ALSO
contain text like "Page 1" or even a literal "[PAGE 1]" in its content. So we
must split on the *real* TTA separators only.

Robustness strategy (all heuristics, no external deps):
  1. The TTA marker is bracketed and sits effectively on its own line:
     ^\\s*\\[\\s*PAGE\\s+N\\s*\\]\\s*$   (case-insensitive)
  2. We accept a marker only if it keeps a (mostly) increasing sequence
     1,2,3,... Stray in-body "[PAGE 1]" that breaks monotonicity is ignored
     and folded into the current page's text instead of starting a new page.
  3. Loose, non-bracketed "Page 1 of 5" is NEVER treated as a separator
     (that's document content, handled later by the page-number layer).

Configurable marker regex so you can match your exact TTA format.
Pure stdlib.
"""
from __future__ import annotations

import re
from typing import List, Optional

# Default: a bracketed PAGE marker alone on a line. Tolerates [PAGE 1], [ Page 12 ],
# and common OCR bracket noise. Adjust to your TTA output if it differs.
DEFAULT_MARKER = re.compile(r"(?im)^[\s>*]*[\[\(<]\s*page\s+(\d{1,4})\s*[\]\)>]\s*$")


def split_pages(ocr_text: str, marker: re.Pattern = DEFAULT_MARKER,
                enforce_sequence: bool = True) -> List[str]:
    """Return ordered per-page text. If no markers are found, returns [ocr_text]."""
    matches = list(marker.finditer(ocr_text))
    if not matches:
        return [ocr_text] if ocr_text.strip() else []

    accepted = _filter_sequence(matches, enforce_sequence)
    if not accepted:
        return [ocr_text]

    pages: List[str] = []
    # text before the first accepted marker (cover text) becomes part of page 1
    first = accepted[0]
    preamble = ocr_text[:first.start()].strip()
    for idx, m in enumerate(accepted):
        body_start = m.end()
        body_end = accepted[idx + 1].start() if idx + 1 < len(accepted) else len(ocr_text)
        body = ocr_text[body_start:body_end].strip("\n")
        if idx == 0 and preamble:
            body = preamble + "\n" + body
        pages.append(body)
    return pages


def _filter_sequence(matches, enforce_sequence: bool):
    """Keep only markers that form a sensible increasing page sequence.

    A marker whose number is not strictly greater than the last accepted one is
    treated as in-document text (the trap) and dropped as a separator.
    """
    if not enforce_sequence:
        return matches
    accepted = []
    last_num: Optional[int] = None
    for m in matches:
        try:
            num = int(m.group(1))
        except (TypeError, ValueError):
            continue
        if last_num is None:
            # accept the first marker only if it starts the sequence low (1 or 2)
            if num <= 2:
                accepted.append(m)
                last_num = num
            # else: likely an in-body "[PAGE 7]" before any real separator -> skip
        else:
            if num > last_num:
                accepted.append(m)
                last_num = num
            # else non-increasing -> in-document text, ignore as a separator
    # Fallback: if sequence filtering rejected everything but markers exist,
    # accept them all (better to over-split than to lose the document).
    return accepted if accepted else matches


def join_pages_with_markers(pages: List[str]) -> str:
    """Inverse helper (used by the generator/tests) — emit TTA-style single string."""
    out = []
    for i, p in enumerate(pages, start=1):
        out.append(f"[PAGE {i}]")
        out.append(p)
    return "\n".join(out)

File: csharp/SegmentationEngine/PageSplitter.cs
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

namespace SegmentationEngine
{
    /// <summary>
    /// Split TTA's single OCR variable into ordered per-page texts.
    /// 1:1 port of python/segmenter/page_splitter.py.
    ///
    /// TTA inserts page separators like [PAGE 1], [PAGE 2], ... A document's own
    /// body can ALSO contain "Page 1" or even a literal "[PAGE 1]" — so we only
    /// accept bracketed markers, alone on a line, that form an increasing sequence.
    /// Loose "Page 1 of 5" is never a separator (handled by the page-number layer).
    /// </summary>
    public static class PageSplitter
    {
        // Bracketed PAGE marker alone on a line. Adjust to your exact TTA format.
        public static readonly Regex DefaultMarker = new Regex(
            @"^[\s>*]*[\[\(<]\s*page\s+(\d{1,4})\s*[\]\)>]\s*$",
            RegexOptions.IgnoreCase | RegexOptions.Multiline | RegexOptions.Compiled);

        public static List<string> SplitPages(string ocrText)
        {
            return SplitPages(ocrText, DefaultMarker, true);
        }

        public static List<string> SplitPages(string ocrText, Regex marker, bool enforceSequence)
        {
            var pages = new List<string>();
            if (string.IsNullOrEmpty(ocrText)) return pages;

            var matches = new List<Match>();
            foreach (Match m in marker.Matches(ocrText)) matches.Add(m);
            if (matches.Count == 0)
            {
                if (ocrText.Trim().Length > 0) pages.Add(ocrText);
                return pages;
            }

            var accepted = FilterSequence(matches, enforceSequence);
            if (accepted.Count == 0) { pages.Add(ocrText); return pages; }

            string preamble = ocrText.Substring(0, accepted[0].Index).Trim();
            for (int idx = 0; idx < accepted.Count; idx++)
            {
                int bodyStart = accepted[idx].Index + accepted[idx].Length;
                int bodyEnd = (idx + 1 < accepted.Count) ? accepted[idx + 1].Index : ocrText.Length;
                string body = ocrText.Substring(bodyStart, bodyEnd - bodyStart).Trim('\n', '\r');
                if (idx == 0 && preamble.Length > 0) body = preamble + "\n" + body;
                pages.Add(body);
            }
            return pages;
        }

        private static List<Match> FilterSequence(List<Match> matches, bool enforce)
        {
            if (!enforce) return matches;
            var accepted = new List<Match>();
            int? last = null;
            foreach (var m in matches)
            {
                int num;
                if (!int.TryParse(m.Groups[1].Value, out num)) continue;
                if (last == null)
                {
                    if (num <= 2) { accepted.Add(m); last = num; }
                }
                else if (num > last.Value)
                {
                    accepted.Add(m); last = num;
                }
            }
            return accepted.Count > 0 ? accepted : matches;
        }

        public static string JoinPagesWithMarkers(List<string> pages)
        {
            var sb = new StringBuilder();
            for (int i = 0; i < pages.Count; i++)
            {
                if (i > 0) sb.Append('\n');
                sb.Append("[PAGE ").Append(i + 1).Append("]\n").Append(pages[i]);
            }
            return sb.ToString();
        }
    }
}
